src/component/gate.rs

```diff
@@ -1,10 +1,41 @@
 use crate::graphics::*;
-use super::gate_definition::GateDefinition;
 
-pub trait Gate<I, O> {
-    fn tick(input: I) -> O;
+pub trait Gate {
+    /// - &mut self: the gate is allowed to change its state when received an input
+    /// - input: list of pointers to inputs in the order they are defined in definition()
+    /// - output: list of pointers to outputs in the order they are defined in definition()
+    fn tick(&mut self, input: GateTickRequest) -> Vec<*const u8>;
 
-    fn draw(direction: Direction, size: Vec2) -> Graphic;
+    /// Produce a graphic of the gate
+    // request will be a reference, because the same gate will be drawn many times
+    // with the same request
+    fn draw(&self, request: &GateDrawRequest) -> Graphic;
 
-    fn definition() -> GateDefinition;
+    // DEFINITION: Returns gate definition
+    // This takes a self reference because the definition may depend on options (which is stored
+    // as state)
+    // below are possible implementations
+    // 1. allow access to a self reference, but this will also allow it to access its state
+    //    but definition should not depend on state
+    // fn definition(&self) -> GateDefinition;
+    //
+    // 2. definition should only depend on the options, so pass in only the option as bytes and do
+    //    the transmut trick on it again
+    // fn definition(option_bytes: *const u8) -> GateDefinition;
```

Comment:
Problem with the second option: it's not dyn-compatible (as it has no `self` argument) meaning that best-case scenario `.definition()` doesn't work on `Box<dyn Gate>`, worst-case-scenario: `dyn Gate` isn't allowed at all.

My proposed solution:
Do the first option but use macros to only give the options to the implementation i.e.
```rust
#[gate]
impl Gate for NandGate {
    pub fn definition(options: ...) {
        // ...
    }

    // transformed to

    pub fn definition(&self) {
        let options = get_options();
        // ... main body, only allowed to use options
        return ...;
    }
} 
```

(I'll figure out the precise types later?)